In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pymongo import MongoClient
from mongoengine import connect
from pandas.io.json import json_normalize
%matplotlib inline

## 如果采用直接导入CSV的形式，请直接跳过这一步骤

In [ ]:
conn = MongoClient(host='127.0.0.1', port=27017,connect=False)  # 实例化MongoClient
db = conn.get_database('maoyan')  # 连接到maoyan数据库

maoyan = db.get_collection('maoyan') # 连接到集合maoyan
mon_data = maoyan.find()  # 查询这个集合下的所有记录，得到的数据类型：pymongo.cursor.Cursor

In [ ]:
type(mon_data)

In [ ]:
#data = json_normalize([comment for comment in mon_data])
#注意这一句需要替换成下面的一句，否则报错
data = pd.json_normalize([comment for comment in mon_data])#得到的数据类型：pandas.core.frame.DataFrame

In [ ]:
type(data)

报错内容：
```Text
<ipython-input-3-de0e8b2488e3>:1: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  data = json_normalize([comment for comment in mon_data])
```
解决方案：
    按照提示将模块语句替换即可

## 如果采用直接导入CSV的形式，请直接跳到这一步骤开始执行

In [5]:
#采用直接读取 CSV 文件的方式将数据读入到 DataFrame 类型变量 data
import csv
data = pd.read_csv("D:\Program Code\Python\Jupyter\课程设计\PythonClassDesignWork\data.csv",delimiter=',', encoding='utf8')

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102584 entries, 0 to 102583
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   content     102583 non-null  object 
 1   gender      102580 non-null  float64
 2   id          102580 non-null  float64
 3   nick        102579 non-null  object 
 4   replyCount  102580 non-null  float64
 5   score       102580 non-null  float64
 6   time        102580 non-null  object 
 7   upCount     102580 non-null  float64
 8   userId      102580 non-null  float64
 9   userLevel   102580 non-null  float64
dtypes: float64(7), object(3)
memory usage: 7.8+ MB


In [7]:
data.head()

content  gender            id  \
0                                            很好震撼的很啊     1.0  1.056560e+09   
1  国产的科幻巅峰，丝毫不逊色好莱坞科幻大片，影响力就像星球大战一样，是国产科幻的开山之做，希望...     0.0  1.056560e+09   
2          超级好看啊 户口帅的一批 我自己一个人看的 看得热血沸腾啊 爷爷死的时候真的哭死了     2.0  1.056558e+09   
3                                         保护生态环境，从我做     1.0  1.056558e+09   
4                                       超级好看，感动，特效好哒     2.0  1.056560e+09   

        nick  replyCount  score                 time  upCount        userId  \
0         江生         0.0    9.0  2019-02-11 10:10:00      0.0  2.016950e+08   
1  Decadence         0.0   10.0  2019-02-11 10:10:00      0.0  1.020876e+09   
2    暴走萝莉912         0.0   10.0  2019-02-11 10:10:00      0.0  3.042996e+08   
3  *^_^*红尘逍遥         0.0   10.0  2019-02-11 10:10:00      0.0  1.548123e+09   
4        罗大大         0.0    9.0  2019-02-11 10:10:00      0.0  8.973641e+08   

   userLevel  
0        3.0  
1        4.0  
2        3.0  
3        2.0  
4        2.0

数据类型转换参考：

| 字段 | 数据类型 |
|------|---------|
| content | object |
| gender | int64 |
| id | int64 |
| nick | object |
| replyCount | int64 |
| score | int64 |
| time | int64 |
| upCount | int64 |
| userId | int64 |
| userLevel | int64 |

In [ ]:
#数据类型转换操作
data['gender'].astype('int64')
data['id'].astype('int64')
data['replyCount'].astype('int64')
data['score'].astype('int64')
data['time'].astype('int64')
data['upCount'].astype('int64')
data['userId'].astype('int64')
data['userLevel'].astype('int64')

In [12]:
# 数据清洗
#data = data.drop(columns='__id') #这一列已经没了，不需要删除丢弃
data = data.drop_duplicates(subset='userId')#去重
data['time'] = pd.to_datetime(data['time']/1000, unit='s')#时间转换，在此之前你必须将表中时间的类型进行转换
data = data[data['time']>=pd.to_datetime('2019-02-05 00:00:00')]
data.set_index(data["time"], inplace=True)
data.head()

TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102584 entries, 0 to 102583
Data columns (total 10 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   content     102583 non-null  object 
 1   gender      102580 non-null  float64
 2   id          102580 non-null  float64
 3   nick        102579 non-null  object 
 4   replyCount  102580 non-null  float64
 5   score       102580 non-null  float64
 6   time        102580 non-null  object 
 7   upCount     102580 non-null  float64
 8   userId      102580 non-null  float64
 9   userLevel   102580 non-null  float64
dtypes: float64(7), object(3)
memory usage: 7.8+ MB


#### 一、数据详情
- 共有102580条数据；
- 包含字段：评论内容、性别、评论ID、评论者昵称、回复数量、评分、时间、点赞数量、评论者ID、评论者等级

#### 二、问题
1. 总体评价如何？
2. 总体评价的时间走向如何？
3. 高分的评价理由是什么？
4. 低分的评价理由是什么？
5. 低分的人群有哪些特征？（性别、等级）
6. 低分跟哪位演员有关？

##### 1. 总体评价如何？

In [ ]:
data['score'].mean()

In [ ]:
from pyecharts import Bar
from pyecharts import Line
from pyecharts import Overlap

score_total = data['score'].value_counts().sort_index()
bar = Bar("《流浪地球》各评分数量", width=700)
line = Line("", width=700)
bar.add("", score_total.index, score_total.values, is_stack=True, is_label_show=True,
       bar_category_gap='40%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
line.add("", score_total.index, score_total.values+1000, is_smooth=True)

overlap = Overlap(width=700)
overlap.add(bar)
overlap.add(line)

overlap

注意使用报错，在这里需要利用pyecharts模块，建议安装0.5.5版本，具体安装命令：
pip install pyecharts==0.5.5 -i https://pypi.douban.com/simple
在新版本中移除了很多东西，会导致模块中部分内容不支持

In [ ]:
# 低分百分比
score_total[:5].sum()/score_total.sum()*100

In [ ]:
# 高分百分比
score_total[7:].sum()/score_total.sum()*100

In [ ]:
# 满分百分比
score_total[10:].sum()/score_total.sum()*100

##### 2. 总体评价的时间走向如何？

In [ ]:
score_by_time = data['score'].resample('H').mean()
line = Line("《流浪地球》平均评分时间走向", width=700)
line.add("", score_by_time.index.date, score_by_time.values, is_smooth=True,
         legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18,
         xaxis_rotate=20, yaxis_min=8)
line

In [ ]:
score_by_time.nsmallest(6)

##### 3. 高分的评价理由是什么？

In [ ]:
import jieba
from collections import Counter
from pyecharts import WordCloud

jieba.add_word('屈楚萧')
jieba.add_word('刘启')
jieba.add_word('吴京')
jieba.add_word('刘培强')
jieba.add_word('李光洁')
jieba.add_word('王磊')
jieba.add_word('吴孟达')
jieba.add_word('达叔')
jieba.add_word('韩子昂')
jieba.add_word('赵今麦')
jieba.add_word('韩朵朵')

swords = [x.strip() for x in open ('stopwords.txt',encoding='utf-8')]

## 上方打开文件需要手动添加编码格式UTF-8

In [ ]:
def plot_word_cloud(data, swords):
    text = ''.join(data['content'])
    words = list(jieba.cut(text))
    ex_sw_words = []
    for word in words:
        if len(word)>1 and (word not in swords):
            ex_sw_words.append(word)
    c = Counter()
    c = Counter(ex_sw_words)
    wc_data = pd.DataFrame({'word':list(c.keys()), 'counts':list(c.values())}).sort_values(by='counts', ascending=False).head(100)
    wordcloud = WordCloud(width=1300, height=620)
    wordcloud.add("", wc_data['word'], wc_data['counts'], word_size_range=[20, 100])
    return wordcloud

In [ ]:
# 高分的评价
plot_word_cloud(data=data[data['score']>6], swords=swords)

In [ ]:
data[data['score']>6].nlargest(10, 'upCount')

In [ ]:
for i in data[data['score']>6].nlargest(10, 'upCount')['content']:
    print(i+'\n')

##### 4. 低分的评价理由是什么？

In [ ]:
# 低分的评价
plot_word_cloud(data=data[data['score']<5], swords=swords)

In [ ]:
data[data['score']<5].nlargest(10, 'upCount')

In [ ]:
for i in data[data['score']<5].nlargest(10, 'upCount')['content']:
    print(i+'\n')

##### 5. 低分的人群有哪些特征？（性别、等级）

In [ ]:
# 总体的性别比例
gender_total = data['gender'].value_counts()
bar = Bar("《流浪地球》观众性别", width=700)
bar.add("", ['未知', '男', '女'], gender_total.values, is_stack=True, is_label_show=True,
       bar_category_gap='60%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

In [ ]:
gender_total/gender_total.sum()*100

In [ ]:
# 低分的性别比例
gender_low = data.loc[data['score']<5, 'gender'].value_counts()
bar = Bar("《流浪地球》低分评论观众性别", width=700)
bar.add("", ['未知', '男', '女'], gender_low.values, is_stack=True, is_label_show=True,
       bar_category_gap='60%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

>提示：
Bar.add用于参数自动补全
Bar.add(self, * args, ** kwargs)

In [ ]:
gender_low/gender_low.sum()*100

In [ ]:
# 总体的等级比例
level_total = data['userLevel'].value_counts().sort_index()
bar = Bar("《流浪地球》观众等级", width=700)
bar.add("", level_total.index, level_total.values, is_stack=True, is_label_show=True,
       bar_category_gap='40%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

In [ ]:
# 低分评论的观众等级比例
level_low = data.loc[data['score']<5, 'userLevel'].value_counts().sort_index()
bar = Bar("《流浪地球》低分评论的观众等级", width=700)
bar.add("", level_low.index, level_low.values, is_stack=True, is_label_show=True,
       bar_category_gap='40%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

## 注意上方报错，导致下方部分数据报错

##### 6. 高低分跟哪位演员有关？

In [ ]:
mapping = {'liucixin':'刘慈欣|大刘', 'guofan':'郭帆', 'quchuxiao':'屈楚萧|刘启|户口', 'wujing':'吴京|刘培强', 
           'liguangjie':'李光洁|王磊', 'wumengda':'吴孟达|达叔|韩子昂', 'zhaojinmai':'赵今麦|韩朵朵'}
for key, value in mapping.items():
    data[key] = data['content'].str.contains(value)

In [ ]:
# 总体提及次数
staff_count = pd.Series({key: data.loc[data[key], 'score'].count() for key in mapping.keys()}).sort_values()
staff_count

In [ ]:
bar = Bar("《流浪地球》演职员总体提及次数", width=700)
bar.add("", ['李光洁','郭帆','赵今麦','吴孟达','屈楚萧','刘慈欣','吴京'], staff_count.values, is_stack=True, is_label_show=True,
       bar_category_gap='60%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

In [ ]:
average_score = pd.Series({key: data.loc[data[key], 'score'].mean() for key in mapping.keys()}).sort_values()
average_score

In [ ]:
bar = Bar("《流浪地球》演职员平均分", width=700)
bar.add("", ['赵今麦','吴孟达','吴京','屈楚萧','李光洁','刘慈欣','郭帆'], np.round(average_score.values,2), is_stack=True, is_label_show=True,
       bar_category_gap='60%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

In [ ]:
staff_count_low = pd.Series({key: data.loc[data[key]&(data['score']<5), 'score'].count() for key in mapping.keys()}).sort_values()
staff_count_low

In [ ]:
staff_count_pct = np.round(staff_count_low/staff_count*100, 2).sort_values()
staff_count_pct

In [ ]:
bar = Bar("《流浪地球》演职员低分评论提及百分比", width=700)
bar.add("", ['郭帆','刘慈欣','李光洁','屈楚萧','赵今麦','吴京','吴孟达'], staff_count_pct.values, is_stack=True, is_label_show=True,
       bar_category_gap='60%', label_color = ['#130f40'],
       legend_text_size=18,xaxis_label_textsize=18,yaxis_label_textsize=18)
bar

In [ ]:
data[data['wumengda']&(data['score']<5)].nlargest(5, 'upCount')

In [ ]:
for i in data[data['wumengda']&(data['score']<5)].nlargest(5, 'upCount')['content']:
    print(i+'\n')

In [ ]:
data[data['wujing']&(data['score']<5)].nlargest(5, 'upCount')

In [ ]:
for i in data[data['wujing']&(data['score']<5)].nlargest(5, 'upCount')['content']:
    print(i+'\n')

In [ ]:
data[data['zhaojinmai']&(data['score']<5)].nlargest(5, 'upCount')

In [ ]:
for i in data[data['zhaojinmai']&(data['score']<5)].nlargest(5, 'upCount')['content']:
    print(i+'\n')